paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/FlinkJobRecoveryITCase.java

```diff
@@ -0,0 +1,320 @@
+                return;
+            case BUCKET_UNAWARE:
+                batchSql(
+                        String.format(
+                                "CREATE TABLE IF NOT EXISTS `%s` (k INT, f1 STRING, pt STRING) WITH ('bucket'='-1', 'commit.force-create-snapshot'='true')",
+                                tableName));
+                return;
+            default:
+                throw new IllegalArgumentException("Unsupported bucket mode: " + bucketMode);
+        }
+    }
+
+    private Snapshot waitForNewSnapshot(String tableName, long initialSnapshot)
+            throws InterruptedException {
+        Snapshot snapshot = findLatestSnapshot(tableName);
+        while (snapshot == null || snapshot.id() == initialSnapshot) {
+            Thread.sleep(2000L);
+            snapshot = findLatestSnapshot(tableName);
+        }
+        return snapshot;
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> T reflectGetField(Object instance, String fieldName)
+            throws NoSuchFieldException, IllegalAccessException {
+        Field field = instance.getClass().getDeclaredField(fieldName);
+        field.setAccessible(true);
+        return (T) field.get(instance);
+    }
+
+    private String triggerCheckpointAndWaitForWrites(JobClient jobClient, long totalReocrds)
+            throws Exception {
+        MiniCluster miniCluster = reflectGetField(jobClient, "miniCluster");
+        JobID jobID = jobClient.getJobID();
+        JobStatus jobStatus = jobClient.getJobStatus().get();
+        while (jobStatus == JobStatus.INITIALIZING || jobStatus == JobStatus.CREATED) {
+            Thread.sleep(2000L);
+            jobStatus = jobClient.getJobStatus().get();
+        }
+
+        if (jobStatus != JobStatus.RUNNING) {
+            throw new IllegalStateException("Job status is not RUNNING");
+        }
+
+        AtomicBoolean allTaskRunning = new AtomicBoolean(false);
+        while (!allTaskRunning.get()) {
+            allTaskRunning.set(true);
+            Thread.sleep(2000L);
+            miniCluster
+                    .getExecutionGraph(jobID)
+                    .thenAccept(
+                            eg ->
+                                    eg.getAllExecutionVertices()
+                                            .forEach(
+                                                    ev -> {
+                                                        if (ev.getExecutionState()
+                                                                != ExecutionState.RUNNING) {
+                                                            allTaskRunning.set(false);
+                                                        }
+                                                    }))
+                    .get();
+        }
+
+        String checkpointPath = miniCluster.triggerCheckpoint(jobID).get();
+        Snapshot snapshot = waitForNewSnapshot("target_table", -1L);
+        while (snapshot.totalRecordCount() < totalReocrds) {
```

Comment:
Unboxing of 'snapshot.totalRecordCount()' may produce 'NullPointerException'

Reply:
Starting from paimon 0.4, `totalRecordCount` is no longer null. So I added `noinspection DataFlowIssue` here to avoid warnings.